### DISCUSSION TITLE
"Hidden" preferences and effective update to workspace?

### DISCUSSION THREAD

**Discussion by VolatileLongPseudoRandom** [eclipse-oomph/oomph#92] [Q&A] [answered]
Hello,

There are a lot of preferences that are not getting "updated" or "triggered" by Oomph preference recorder or in the other way (from Oomph to the Workspace), such as:

- The layout of the Problem view (`org.eclipse.ui.ide`/`org.eclipse.ui.internal.views.markers.CachedMarkerBuilderorg.eclipse.ui.views.ProblemView`)
- The do not display on stdout/stderr in the Console (`org.eclipse.debug.ui`/`DEBUG.consoleOpenOnOut` or `DEBUG.consoleOpenOnErr`)
- Errors logs to group by session (`org.eclipse.ui.views.log`/`groupBy`=`1`) and its activation (`org.eclipse.ui.views.log`/`activate`=`false`)
- Default perspective for launch mode (`org.eclipse.debug.ui`/`org.eclipse.debug.ui.PREF_LAUNCH_PERSPECTIVES`) (which also require a plugin contributing a perspective)
- Toolbar for debug (`org.eclipse.debug.ui`/`org.eclispe.debug.ui.Debug_view.debug_toolbar_hidden_perspectives`)

I've found most of these using the _Preferences Spy_ even if in some case (like Error log _Activate on new events_) it did take some time before preference was found modified.

Now I am trying to understand how I can synchronize:

- from `user.setup` to workspace without restarting the workspace
- from workspace to `user.setup`.

For example, if I change the `activate` of the `Error logs` in one Workspace, and open another workspace (it could be another instance of Eclipse, let's say I worked on Project A then I work on project B): the preference is updated:

```log
Performing Preference /instance/org.eclipse.ui.views.log/activate = true
```

But the Error log _Activate on new events_ is not checked and it seems, I need to restart the workspace: is this a limitation of Oomph or in Eclipse own Preference mechanism? (eg: if one plugin update preference of another, the plugin never receive the appropriate change event?)

In the other way, if in the `user.setup` _Activate on new events_ is true and I want to disable it, the recorder never record it and I did not find any way to force it outside the Preferences window.

Also, because I store the `user.setup` in my private repositories as backup, is there a way to change the Oomph XML to use a line length of 150 characters?

**Comment by merks** [comment, marked as answer]
There are a lot of question here....   

One quite useful tool that is not installed by default is the preference manager:
```
<?xml version="1.0" encoding="UTF-8"?>
<setup.p2:P2Task
    xmi:version="2.0"
    xmlns:xmi="http://www.omg.org/XMI"
    xmlns:setup.p2="http://www.eclipse.org/oomph/setup/p2/1.0">
  <requirement
      name="org.eclipse.oomph.preferences.feature.group"/>
</setup.p2:P2Task>
```
This provides the following:

![image](https://github.com/eclipse-oomph/oomph/assets/208716/dc7de6f0-927f-45ed-9ddd-42884731ae9b)

I.e., it shows a tree of all the preferences.  You can open it in an editor:

![image](https://github.com/eclipse-oomph/oomph/assets/208716/64009482-ec3c-4e11-b7a7-a90c4d5a3555)

With the sync button enabled, it will select preferences as they are changed by something else in IDE.

There's also this button to capture preferences so you can covert them to preference tasks in a setup:

![image](https://github.com/eclipse-oomph/oomph/assets/208716/f3702c09-b4aa-4d23-99aa-571338b88516)

I'm not sure all the things you describe are preferences.  Some things are recorded in dialog settings. This is all I see for org.eclipse.ui.ide at least under instance preferences:

![image](https://github.com/eclipse-oomph/oomph/assets/208716/c236b534-de01-41fe-af98-2c033178422c)

It may well be the case that some settings are only recognized as changed if one uses the UI provide by the part of the framework using that preference, i.e., not all parts of the framework look at the preference each time and not all parts listen to preference changes.  The Preference Task can't know what the downstream frameworks are doing with any specific preference value in the preference tree.  Each case likely has a different answer...

The line width is hard coded:

![image](https://github.com/eclipse-oomph/oomph/assets/208716/f653866c-13c8-44bb-aa4e-d4b7c05a5e76)

I would have expected that to have each attribute on a separate line helps reduce the confusing noise when doing a textual comparison.  I suppose we could use a system property to change this value, or a preference, but that's much more complex.

**Reply by VolatileLongPseudoRandom** [reply]
Thanks for the reply, I'll try to give the Preference manager a try especially since I've observed that the Eclipse 2024.06 added some new preference for the Error log ... :)

For the indentation part, that's probably because I'm accustomed to having an XML editor with 150 character per line : I prefer editing the file using vscode rather than the visual editor, especially because it get very slow at some time (and I think it does not free resources, because my Eclipse had its 2G memory fully consumed even after closing all setup file).
